# Git Hub Project #2 Data Cleaning with Text in Pandas

### Importing Packages

In [1]:
import os
import sys
import numpy as np
import pandas as pd
import re
from pathlib import Path

## Importing Data

In [2]:
Path = r"/Users/peterlyon/Downloads/Coding+Test+L3+Datafiles/patent_drawing.csv"
Patents = pd.read_csv(os.path.normpath(Path))
Patents

uuid patent_id  \
0     urfl2ulyjgez01g5selfflnz7   4491930   
1     nqdxwthotlcted3d961ao373x   4490979   
2     9mwinm7as0p0j3245tdxhfuiz   4491969   
3     l1n6w0ofqic6yow2t7qwmvqry   4490948   
4     86bndneq4omf3mfxi60dzr5mx   4491426   
...                         ...       ...   
8151  bs3gaqy956cbznfcscxl8zrrj   4491812   
8152  nyj0iok0vet86cdkt0h70f53r   4491485   
8153  wpvtafbht8yhm3lcvlc53r7hv   4491167   
8154  jwmtihlklmnjnjaok71ngr8ia   4490920   
8155  f6elpj5w2b4kbvv46dyt76dt8   4491024   

                                                   text  
0     A better understanding of the invention may be...  
1     A better understanding of the invention will b...  
2     A better understanding of the invention will b...  
3     A better understanding of the present inventio...  
4     A better understanding of the present inventio...  
...                                                 ...  
8151  While the specification concludes with claims ...  
8152  With particular reference to FIG. 2, it is not...  
8153  With the above and other objects and advantage...  
8154  With the foregoing and other objects in view, ...  
8155  Yet further features and advantages of the inv...  

[8156 rows x 3 columns]

In [3]:
Patents.describe()

uuid patent_id                              text
count                        8156      8156                              8156
unique                       8156      1096                              7722
top     urfl2ulyjgez01g5selfflnz7   4491287  BRIEF DESCRIPTION OF THE DRAWING
freq                            1        59                               120

## Question 1

How many of the descriptions mention an embodiment or that they embody something? That is, in how many of the description does the stem “embod-” appear with any ending? (i.e., embody, embodiment, embodying, etc. should all count).

In [4]:
Patents.loc[(Patents["text"].str.contains(r'embod', case=False, regex=True))]

uuid patent_id  \
6     hi5rsi8g5wx87octi0c6yqzfs   4491821   
7     0pw76mclimgppe77x6lzmfoa1   4491934   
13    kowiz2l3dzwexqfik80zyk3dh   D276953   
24    aij6voi71zrp8flvp50ng5zjn   4491267   
25    6frag6oj4fnsef5yz99cf177d   4491019   
...                         ...       ...   
8140  yn4e3eaegufv7wfvun77zccp8   4491687   
8141  b5eu7hz1hz0alu358kxk8arg2   4491066   
8148  f1ueevkbtxs9n4x5tebzbkyo9   4491045   
8153  wpvtafbht8yhm3lcvlc53r7hv   4491167   
8154  jwmtihlklmnjnjaok71ngr8ia   4490920   

                                                   text  
6     A better understanding of the present inventio...  
7     A better understanding of the present inventio...  
13    A front perspective view with the front top fl...  
24    A preferred embodiment is seen in the attached...  
25    A preferred embodiment of the invention is des...  
...                                                 ...  
8140  Two embodiments of the invention are described...  
8141  Two embodiments of the invention will now be d...  
8148  While the specification concludes with claims ...  
8153  With the above and other objects and advantage...  
8154  With the foregoing and other objects in view, ...  

[1317 rows x 3 columns]

In [5]:
Patents["Embody"] = Patents["text"].str.contains(r'embod', case=False, regex=True)

In [6]:
Patents.head()

uuid patent_id  \
0  urfl2ulyjgez01g5selfflnz7   4491930   
1  nqdxwthotlcted3d961ao373x   4490979   
2  9mwinm7as0p0j3245tdxhfuiz   4491969   
3  l1n6w0ofqic6yow2t7qwmvqry   4490948   
4  86bndneq4omf3mfxi60dzr5mx   4491426   

                                                text  Embody  
0  A better understanding of the invention may be...   False  
1  A better understanding of the invention will b...   False  
2  A better understanding of the invention will b...   False  
3  A better understanding of the present inventio...   False  
4  A better understanding of the present inventio...   False

In [7]:
Patents["Embody"].value_counts()

False    6839
True     1317
Name: Embody, dtype: int64

## Question 1 Answer

1317 Cases of “embod-” appear with various endings.

## Question 2

What if we are specifically interested in drawing descriptions that embody an invention? So, here, we would like to identify descriptions that contain phrases like the &quot;embodiment of the present invention, "embodies my invention", etc. How many descriptions have the word  “embod-” (with any ending) followed by "invention", even if the two terms are separated by words?

In [8]:
Patents.loc[(Patents["text"].str.contains('invention', case=False, regex=True))]

uuid patent_id  \
0     urfl2ulyjgez01g5selfflnz7   4491930   
1     nqdxwthotlcted3d961ao373x   4490979   
2     9mwinm7as0p0j3245tdxhfuiz   4491969   
3     l1n6w0ofqic6yow2t7qwmvqry   4490948   
4     86bndneq4omf3mfxi60dzr5mx   4491426   
...                         ...       ...   
8150  es2msvdn5pd8zxnxwrsmb3frd   4491306   
8151  bs3gaqy956cbznfcscxl8zrrj   4491812   
8153  wpvtafbht8yhm3lcvlc53r7hv   4491167   
8154  jwmtihlklmnjnjaok71ngr8ia   4490920   
8155  f6elpj5w2b4kbvv46dyt76dt8   4491024   

                                                   text  Embody  
0     A better understanding of the invention may be...   False  
1     A better understanding of the invention will b...   False  
2     A better understanding of the invention will b...   False  
3     A better understanding of the present inventio...   False  
4     A better understanding of the present inventio...   False  
...                                                 ...     ...  
8150  While the specification concludes with claims ...   False  
8151  While the specification concludes with claims ...   False  
8153  With the above and other objects and advantage...    True  
8154  With the foregoing and other objects in view, ...    True  
8155  Yet further features and advantages of the inv...   False  

[2297 rows x 4 columns]

In [9]:
Patents["invention"] = Patents["text"].str.contains('invention', case=False, regex=True)

In [10]:
Patents.head()

uuid patent_id  \
0  urfl2ulyjgez01g5selfflnz7   4491930   
1  nqdxwthotlcted3d961ao373x   4490979   
2  9mwinm7as0p0j3245tdxhfuiz   4491969   
3  l1n6w0ofqic6yow2t7qwmvqry   4490948   
4  86bndneq4omf3mfxi60dzr5mx   4491426   

                                                text  Embody  invention  
0  A better understanding of the invention may be...   False       True  
1  A better understanding of the invention will b...   False       True  
2  A better understanding of the invention will b...   False       True  
3  A better understanding of the present inventio...   False       True  
4  A better understanding of the present inventio...   False       True

In [11]:
Patentsv1 = Patents[Patents["Embody"] & Patents["invention"]]
Patentsv1.head()

uuid patent_id  \
6   hi5rsi8g5wx87octi0c6yqzfs   4491821   
7   0pw76mclimgppe77x6lzmfoa1   4491934   
25  6frag6oj4fnsef5yz99cf177d   4491019   
26  iwrna7u45dat2kj8clsbwct5u   4491391   
27  fo6110e56ijjpu5f5x2jygqih   4491418   

                                                 text  Embody  invention  
6   A better understanding of the present inventio...    True       True  
7   A better understanding of the present inventio...    True       True  
25  A preferred embodiment of the invention is des...    True       True  
26  A preferred embodiment of the invention is sho...    True       True  
27  A preferred embodiment of the invention will n...    True       True

In [12]:
for text in Patents["text"]:
    if r"embod" in text and "invention" in text and text.index(r"embod") < text.index("invention"):
        print(text)

A preferred embodiment of the invention is described hereinafter with reference to the accompanying drawing wherein
A preferred embodiment of the invention is shown in the figures of the drawing wherein:
A preferred embodiment of the invention will now be described in detail with reference to the accompanying drawing, in which
A preferred embodiment of the invention will now be described in detail with reference to the accompanying drawing, which is a partial perspective view showing the spray head unscrewed and moved slightly away from the pressurized-water supply line for the sake of clarity.
A preferred embodiment of the present invention will now be described with reference to the example thereof as illustrated in the accompanying drawings, in which:
A specific embodiment of the present invention will now be described in detail, by way of example only, with reference to the accompanying drawings, in which;
A specific embodiment of the present invention, showing our present contempl

In [13]:
def is_embod_and_invention(text):
    return bool(re.search(r"embod", text) and re.search(r"invention", text) and
                re.search(r"embod", text).start() < re.search(r"invention", text).start())

Patents["is_embod_then_invention"] = Patents["text"].apply(is_embod_and_invention)
Patents

uuid patent_id  \
0     urfl2ulyjgez01g5selfflnz7   4491930   
1     nqdxwthotlcted3d961ao373x   4490979   
2     9mwinm7as0p0j3245tdxhfuiz   4491969   
3     l1n6w0ofqic6yow2t7qwmvqry   4490948   
4     86bndneq4omf3mfxi60dzr5mx   4491426   
...                         ...       ...   
8151  bs3gaqy956cbznfcscxl8zrrj   4491812   
8152  nyj0iok0vet86cdkt0h70f53r   4491485   
8153  wpvtafbht8yhm3lcvlc53r7hv   4491167   
8154  jwmtihlklmnjnjaok71ngr8ia   4490920   
8155  f6elpj5w2b4kbvv46dyt76dt8   4491024   

                                                   text  Embody  invention  \
0     A better understanding of the invention may be...   False       True   
1     A better understanding of the invention will b...   False       True   
2     A better understanding of the invention will b...   False       True   
3     A better understanding of the present inventio...   False       True   
4     A better understanding of the present inventio...   False       True   
...                                                 ...     ...        ...   
8151  While the specification concludes with claims ...   False       True   
8152  With particular reference to FIG. 2, it is not...   False      False   
8153  With the above and other objects and advantage...    True       True   
8154  With the foregoing and other objects in view, ...    True       True   
8155  Yet further features and advantages of the inv...   False       True   

      is_embod_then_invention  
0                       False  
1                       False  
2                       False  
3                       False  
4                       False  
...                       ...  
8151                    False  
8152                    False  
8153                    False  
8154                    False  
8155                    False  

[8156 rows x 6 columns]

In [14]:
Patents.describe()

uuid patent_id                              text  \
count                        8156      8156                              8156   
unique                       8156      1096                              7722   
top     urfl2ulyjgez01g5selfflnz7   4491287  BRIEF DESCRIPTION OF THE DRAWING   
freq                            1        59                               120   

       Embody invention is_embod_then_invention  
count    8156      8156                    8156  
unique      2         2                       2  
top     False     False                   False  
freq     6839      5859                    7386

In [15]:
Patents["is_embod_then_invention"].value_counts()

False    7386
True      770
Name: is_embod_then_invention, dtype: int64

## Question 2 Answer

770 Cases of “embod-” appear with various endings with invention appearing afterword.